src/main/host/syscall_condition.c

```diff
@@ -39,22 +38,32 @@ struct _SysCallCondition {
     MAGIC_DECLARE;
 };
 
-SysCallCondition* syscallcondition_new(Timer* timeout, Descriptor* desc,
-                                       DescriptorStatus status) {
+SysCallCondition* syscallcondition_new(Trigger trigger, Timer* timeout) {
     SysCallCondition* cond = malloc(sizeof(*cond));
 
-    *cond = (SysCallCondition){.timeout = timeout,
-                               .desc = desc,
-                               .status = status,
-                               .referenceCount = 1,
-                               MAGIC_INITIALIZER};
+    *cond = (SysCallCondition){
+        .timeout = timeout, .trigger = trigger, .referenceCount = 1, MAGIC_INITIALIZER};
 
     /* We now hold refs to these objects. */
-    if (timeout) {
-        descriptor_ref(timeout);
+    if (cond->timeout) {
+        descriptor_ref(cond->timeout);
     }
-    if (desc) {
-        descriptor_ref(desc);
+
+    if (cond->trigger.object.as_pointer) {
+        switch (cond->trigger.type) {
+            case TRIGGER_DESCRIPTOR: {
+                descriptor_ref(cond->trigger.object.as_descriptor);
+                break;
+            }
+            case TRIGGER_FUTEX: {
+                futex_ref(cond->trigger.object.as_futex);
+                break;
+            }
+            case TRIGGER_NONE:
+            default: {
```

Comment:
warning in default case?

Reply:
alternatively no `default` case at all to get a compiler warning if another enumerator gets added later without updating this switch; `return` statements in each case; and a `warning` or `error` after the whole statement to detect at runtime if we got a bad enumerator here.

Blatant self-promotion (a bit Google-centric though): https://abseil.io/tips/147

e.g.
```
        switch (cond->trigger.type) {
            case TRIGGER_DESCRIPTOR: {
                descriptor_ref(cond->trigger.object.as_descriptor);
                return;
            }
            case TRIGGER_FUTEX: {
                futex_ref(cond->trigger.object.as_futex);
                return;
            }
            case TRIGGER_NONE: {
                 return;
            }
            // no default, forcing a compiler warning if not kept up to date
        }
        // Complain loudly if we get a non-enumerator at run-time.
        error("Unhandled enumerator %d", cond->trigger.type);
```